### PR TITLE
Add relational let nodes

### PR DESCRIPTION
**Pull request by tpoterba**
Stacked on #5864

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/Parser.scala [hunk 1]
```diff
@@ -584,6 +584,15 @@ object IRParser {
       case "Ref" =>
         val id = identifier(it)
         Ref(id, env.refMap(id))
+      case "PipelineRef" =>
```
Comment (cseed) [inline hail/src/main/scala/is/hail/expr/ir/Parser.scala]: Should be RelationalRef?

File: hail/src/main/scala/is/hail/expr/ir/EvaluateRelationalLets.scala [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package is.hail.expr.ir
+
+import scala.collection.mutable
+
+object EvaluateRelationalLets {
+
+  def apply(ir: BaseIR): BaseIR = {
+
+    def interpretAndCoerce(value: IR): IR = Literal.coerce(value.typ, Interpret(value, optimize = false))
```
Comment (cseed) [inline hail/src/main/scala/is/hail/expr/ir/EvaluateRelationalLets.scala]: We never want to call Interpret, do we?  Don't want CompileAndEvaluate here?

File: hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala [hunk 3]
```diff
@@ -6,7 +6,17 @@ import is.hail.expr.types.virtual._
...
+import scala.collection.mutable
+
+
 object PruneDeadFields {
+
+  case class ComputeMutableState(requestedType: Memo[BaseType], relationalRefs: mutable.HashMap[String, ArrayBuilder[Type]]) {
+    def rebuildState: RebuildMutableState = RebuildMutableState(requestedType, mutable.HashMap.empty)
+  }
+
+  case class RebuildMutableState(requestedType: Memo[BaseType], relationalRefs: mutable.HashMap[String, Type])
```
Comment (cseed) [inline hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala]: These classes have strange names.  They are verbs but classes are normally nouns.  Why is compute have an array of types for compute?

Comment (cseed) [reply hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala]: Ah, is this the mutable state of the compute and rebuild steps, respectively?  Compute here means the memo step?

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala]: yes, exactly.

File: hail/src/main/scala/is/hail/expr/ir/EvaluateRelationalLets.scala (hunk 2 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/EvaluateRelationalLets.scala]: in the case that these things will actually be used in, it'll be a TableAggregate or something, which needs to go through Interpret. 

I'm going to fix up the interpret/compileAndEvaluate stuff soon to never use interpret.

File: hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala (hunk 3 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala]: are you asking about the ArrayBuilder[Type]? These are the usages of the relational refs.

File: hail/src/main/scala/is/hail/expr/ir/EvaluateRelationalLets.scala (hunk 2 above)
Comment (cseed) [reply hail/src/main/scala/is/hail/expr/ir/EvaluateRelationalLets.scala]: > I'm going to fix up the interpret/compileAndEvaluate stuff soon to never use interpret.

Awesome.  All these entrypoints and code paths and recursive invocations gives me a bella ache.  Maybe move Interpret to the test module?

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/EvaluateRelationalLets.scala]: yeah, that's my plan!

**Pull request merged**